### DISCUSSION TITLE
Optimality

### DISCUSSION THREAD

**Discussion by antoniojbm** [aiplan4eu/unified-planning#495] [Q&A]
Good morning, I have been reading documentation about PDDL, I am very interested in the optimization part especially the part of "MinimizeMakespan" and "MinimizeExpressionOnFinalState", I would like to use it in my project but I can't find an example of these two features. Please, could you give me short examples of how they should be used?

**Comment by Framba-Luca** [comment]
Hi @antoniojbm, to use the metrics it's enough to import them, create an instance and add the instance to the problem with `problem.add_quality_metric(metric)`.

- `MinimizeMakespan` doesn't need any argument, so it can be created with `metric = MinimizeMakespan()`; the only problem is that creating this metric is useless, because we don't have any solver that supports it.
- `MinimizeExpressionOnFinalState` takes an expression; this expression must be numeric and can be, for example, a numeric fluent; this metric is supported only by [Enhsp](https://github.com/aiplan4eu/up-enhsp/blob/master/up_enhsp/enhsp_planner.py#L79), that does not support temporal problems. 
 
There are 2 enhsp engines: 
1. [ENHSPEngine](https://github.com/aiplan4eu/up-enhsp/blob/master/up_enhsp/enhsp_planner.py#L18), that supports the metric but does not guarantee the optimality (it might be an optimal plan, the planner takes in consideration the metric, but it's not guaranteed)
2. [ENHSPOptEngine](https://github.com/aiplan4eu/up-enhsp/blob/master/up_enhsp/enhsp_planner.py#L126), that supports the metric, guarantees the optimality, but supports only a smaller range of ProblemKinds (as you can see [here](https://github.com/aiplan4eu/up-enhsp/blob/master/up_enhsp/enhsp_planner.py#L139C3-L139C3))

Looking at the [optimality notebook](https://unified-planning.readthedocs.io/en/latest/notebooks/02-optimal-planning.html) it should be clear how to use those metrics, even if they are different; if you still have problems in using them just let me know.

**Reply by antoniojbm** [reply]
I looked at all the documentation but it was not clear to me how to use it, now everything is very clear. Thank you very much for your quick response and for always explaining everything so well. One last questions: 


- I am working with Tamer, is there any option to optimize the results with this engine? I am interested in optimizing the total time. 

- And i have other question, are PDDL 2.2 Timed Initial Literals are supported?



P.D.: Also comment that in the documentation where it says ""MinimizeMakespan (Temporal planning only) Asks to find a plan taking the shortest possible time until the ending of the last action in the plan." implies that it works with any engine, maybe some annotation remarking which engines are compatible would help a lot.

**Comment by arbimo** [comment]
Aries actually supports the makespan optimization metric but it may not be efficient to your problem (its focus is on non-recursive hierchical planning). Note that it should be used with the anytime operation mode to get solutions of increasing quality even in hte absence of optimality proof (which aries may never give you for certain classes of problems).
